nexus/db-queries/src/transaction_retry.rs

```diff
@@ -0,0 +1,141 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Helper types for performing automatic transaction retries
+
+use chrono::Utc;
+use oximeter::{types::Sample, Metric, MetricsError, Target};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+// Identifies "which" transaction is retrying
+#[derive(Debug, Clone, Target)]
+struct Transaction {
+    name: String,
+}
+
+// Identifies that a retry has occurred, and track how long
+// the transaction took (either since starting, or since the last
+// retry failure was recorded).
+#[derive(Debug, Clone, Metric)]
+struct RetryData {
+    #[datum]
+    latency: f64,
+}
+
+// Collects all transaction retry samples
+#[derive(Debug, Default, Clone)]
+pub(crate) struct Producer {
+    samples: Arc<Mutex<Vec<Sample>>>,
+}
+
+impl Producer {
+    pub(crate) fn new() -> Self {
+        Self { samples: Arc::new(Mutex::new(vec![])) }
+    }
+
+    fn append(
+        &self,
+        transaction: &Transaction,
+        latency: Duration,
+    ) -> Result<(), MetricsError> {
+        let sample = Sample::new_with_timestamp(
+            Utc::now(),
+            transaction,
+            &RetryData { latency: latency.as_secs_f64() },
+        )?;
+        self.samples.lock().unwrap().push(sample);
+        Ok(())
+    }
+}
+
+struct RetryHelperInner {
+    start: chrono::DateTime<Utc>,
+    attempts: usize,
+}
+
+impl RetryHelperInner {
+    fn new() -> Self {
+        Self { start: Utc::now(), attempts: 1 }
+    }
+
+    fn tick(&mut self) -> Self {
+        let start = self.start;
+        let attempts = self.attempts;
+
+        self.start = Utc::now();
+        self.attempts += 1;
+
+        Self { start, attempts }
+    }
+}
+
+/// Helper utility for tracking retry attempts and latency.
+/// Intended to be used from within "transaction_async_with_retry".
+pub(crate) struct RetryHelper {
+    producer: Producer,
+    name: &'static str,
+    inner: Mutex<RetryHelperInner>,
+}
+
+const MAX_RETRY_ATTEMPTS: usize = 10;
+
+impl RetryHelper {
+    /// Creates a new RetryHelper, and starts a timer tracking the transaction
+    /// duration.
+    pub(crate) fn new(producer: &Producer, name: &'static str) -> Self {
+        Self {
+            producer: producer.clone(),
+            name,
+            inner: Mutex::new(RetryHelperInner::new()),
+        }
+    }
+
+    // Called upon retryable transaction failure.
+    //
+    // This function:
+    // - Appends a metric identifying the duration of the transaction operation
+    // - Performs a randomly generated backoff (limited to less than 50 ms)
+    // - Returns "true" if the transaction should be restarted
+    async fn retry_callback(&self) -> bool {
+        let inner = self.inner.lock().unwrap().tick();
+
+        let _ = self.producer.append(
+            &Transaction { name: self.name.into() },
+            (Utc::now() - inner.start).to_std().unwrap_or(Duration::ZERO),
```

Comment:
So if I'm following all this, the start time is taken pretty early, in some cases before we even take a connection from the pool. Is it possible to tighten those bounds at all, say by taking the timestamp just before the transaction closure is actually run? That may not be possible without more coordination in the actual connect pool implementation.

If not, I think it'd help to have a note that the latency is an upper bound, since it includes setup and network time.

Reply:
Yeah, this is definitely an issue here. I figured it wouldn't be *that* bad, because this is basically a developer-facing issue, and we care about:

- Which transactions are open for a long time and hitting retry errors (should they be optimized?)
- Which transactions are open for a short time and hitting retry errors (should they be converted to CTEs?)

But beyond that, I don't really know how much precision matters. I think, by far, the most important aspect is just going to be a ranking of "which transactions are retrying the most"?

I do think it's possible to get a more picture-perfect accurate view, but I don't *think* the accuracy matters enough to do it -- it would require basically making a call to the `RetryHelper` object within the body of the function passed to `transaction_async_with_callback`, which would also require wrapping it in an `Arc` across all usages. I don't *think* the tradeoff is worth it? I'd bet the cost of checking out a connection is small relative to the cost of most transactions.

If you disagree, I can sketch out a more real plan, but given our usage here (dev only visibility for metrics, to help us prioritize) I'm kinda making the bet this shortcut is worth it.

Reply:
I think you're probably right that it's not worth much complexity at this point, but let's add a note that this includes a bit of time that's not strictly part of the transaction in the database itself.